File: vscode-app/src/providers/ProviderProfile.ts
export type ProviderId = 'mistral' | 'ollama' | 'deepseek' | '3dllm';

export interface ProviderModelSettings {
  temperature: number;
  maxTokens: number;
  topP: number;
  stop?: string[];
}

export interface ProviderToolPolicy {
  mode: 'native' | 'prompt';
  promptRules?: string;
}

export interface ProviderPromptPolicy {
  systemRole?: string;
  toolRules?: string;
}

export interface ProviderSessionPolicy {
  compactionThreshold: number;
  resetOnProviderChange: boolean;
}

export interface ProviderProfile {
  id: ProviderId;
  modelDefaults: ProviderModelSettings;
  toolPolicy: ProviderToolPolicy;
  prompt?: ProviderPromptPolicy;
  sessionPolicy: ProviderSessionPolicy;
}

const RESET_SESSION = true;

const TOOL_RULES_COMMON = 'You may call tools using the tool-call protocol described in the system instructions. Use only tools that are currently available. When a tool is required, return exactly one JSON tool call. Do not describe a tool call in prose; emit the tool call itself.';

const TOOL_RULES_3DLLM = 'Tool-call format: { "tool": "toolName", "input": {} }. Always call a tool when a task requires filesystem, terminal, build, or edit operations. If tool output contains an error, fix the input and retry once. After the final tool result, answer the user directly.';

export const PROVIDER_PROFILES: Record<ProviderId, ProviderProfile> = {
  mistral: {
    id: 'mistral',
    modelDefaults: { temperature: 0.7, maxTokens: 4096, topP: 1, stop: [] },
    toolPolicy: { mode: 'prompt', promptRules: TOOL_RULES_COMMON },
    prompt: { toolRules: TOOL_RULES_COMMON },
    sessionPolicy: { compactionThreshold: 0.8, resetOnProviderChange: RESET_SESSION },
  },
  ollama: {
    id: 'ollama',
    modelDefaults: { temperature: 0.8, maxTokens: 2048, topP: 0.9 },
    toolPolicy: { mode: 'prompt', promptRules: TOOL_RULES_COMMON },
    prompt: { toolRules: TOOL_RULES_COMMON },
    sessionPolicy: { compactionThreshold: 0.75, resetOnProviderChange: RESET_SESSION },
  },
  deepseek: {
    id: 'deepseek',
    modelDefaults: { temperature: 0.7, maxTokens: 8192, topP: 1 },
    toolPolicy: { mode: 'prompt', promptRules: TOOL_RULES_COMMON },
    prompt: { toolRules: TOOL_RULES_COMMON },
    sessionPolicy: { compactionThreshold: 0.85, resetOnProviderChange: RESET_SESSION },
  },
  '3dllm': {
    id: '3dllm',
    modelDefaults: { temperature: 0.2, maxTokens: 4096, topP: 0.95 },
    toolPolicy: { mode: 'prompt', promptRules: TOOL_RULES_3DLLM },
    prompt: { toolRules: TOOL_RULES_3DLLM },
    sessionPolicy: { compactionThreshold: 0.9, resetOnProviderChange: RESET_SESSION },
  },
};

export function getProviderProfile(providerId: string): ProviderProfile {
  const key = providerId as ProviderId;
  return PROVIDER_PROFILES[key] ?? PROVIDER_PROFILES.ollama;
}

export function mergeModelSettings(
  defaults: ProviderModelSettings,
  overrides: Partial<ProviderModelSettings> = {},
): ProviderModelSettings {
  return {
    temperature: overrides.temperature ?? defaults.temperature,
    maxTokens: overrides.maxTokens ?? defaults.maxTokens,
    topP: overrides.topP ?? defaults.topP,
    stop: overrides.stop ?? defaults.stop,
  };
}

export function getToolRules(profile: ProviderProfile, toolsEnabled: boolean): string | undefined {
  if (!toolsEnabled) {
    return undefined;
  }
  return profile.prompt?.toolRules ?? profile.toolPolicy.promptRules;
}

File: llm-client/src/main/kotlin/com/i2vision/llm/ProviderProfile.kt
/*
 * Copyright (c) 2026. Oleksii Lykov.
 *
 * Licensed under the MIT License.
 * SPDX-License-Identifier: MIT
 */

package com.i2vision.llm

enum class ToolCallingMode {
    PROMPT_BASED,
    NATIVE
}

data class ProviderCapabilities(
    val supportsStreaming: Boolean = false,
    val toolCallingMode: ToolCallingMode = ToolCallingMode.PROMPT_BASED,
    val supportsVision: Boolean = false,
    val maxContextTokens: Int = 8192
)

data class ModelSettings(
    val temperature: Double = 0.7,
    val topP: Double = 0.9,
    val topK: Int = 40,
    val maxTokens: Int = 2048,
    val stop: List<String> = emptyList()
)

data class ProviderPromptProfile(
    val systemRole: String? = null,
    val toolRules: String? = null
)

data class SessionPolicy(
    val compactionThreshold: Int = 85,
    val resetOnProviderChange: Boolean = true
)

data class ProviderProfile(
    val provider: LLMProvider,
    val modelDefaults: ModelSettings = ModelSettings(),
    val capabilities: ProviderCapabilities = ProviderCapabilities(),
    val prompt: ProviderPromptProfile = ProviderPromptProfile(),
    val sessionPolicy: SessionPolicy = SessionPolicy()
)

data class ProviderResolution(
    val provider: LLMProvider,
    val profile: ProviderProfile
)

object ProviderProfiles {

    fun forProvider(provider: LLMProvider): ProviderProfile = when (provider) {
        LLMProvider.OLLAMA -> ProviderProfile(
            provider = provider,
            modelDefaults = ModelSettings(
                temperature = 0.7,
                topP = 0.9,
                topK = 40,
                maxTokens = 2048
            ),
            capabilities = ProviderCapabilities(
                supportsStreaming = false,
                toolCallingMode = ToolCallingMode.PROMPT_BASED,
                supportsVision = false,
                maxContextTokens = 8192
            ),
            prompt = ProviderPromptProfile(
                systemRole = null,
                toolRules = """
                    Tool rules:
                    - When you need data outside your own knowledge, call a tool.
                    - Return only the JSON object expected by the tool schema.
                    - If the previous tool call failed validation, correct the JSON and retry once.
                    - Never invent tool output; wait for the actual tool response.
                """.trimIndent()
            ),
            sessionPolicy = SessionPolicy(
                compactionThreshold = 85,
                resetOnProviderChange = true
            )
        )

        LLMProvider.DEEPSEEK -> ProviderProfile(
            provider = provider,
            modelDefaults = ModelSettings(
                temperature = 0.7,
                topP = 0.9,
                topK = 40,
                maxTokens = 4096
            ),
            capabilities = ProviderCapabilities(
                supportsStreaming = true,
                toolCallingMode = ToolCallingMode.PROMPT_BASED,
                supportsVision = false,
                maxContextTokens = 65536
            ),
            prompt = ProviderPromptProfile(
                systemRole = null,
                toolRules = """
                    Tool rules:
                    - Use tools only when the requested information cannot be answered directly.
                    - Emit a single valid JSON tool call with required fields only.
                    - If a tool call fails, recover by fixing the input schema.
                """.trimIndent()
            )
        )

        LLMProvider.MISTRAL -> ProviderProfile(
            provider = provider,
            modelDefaults = ModelSettings(
                temperature = 0.7,
                topP = 0.9,
                topK = 40,
                maxTokens = 4096
            ),
            capabilities = ProviderCapabilities(
                supportsStreaming = true,
                toolCallingMode = ToolCallingMode.PROMPT_BASED,
                supportsVision = false,
                maxContextTokens = 32768
            ),
            prompt = ProviderPromptProfile(
                systemRole = null,
                toolRules = """
                    Tool rules:
                    - Prefer native tool calls when available.
                    - Use concise tool inputs and match field names exactly.
                    - Do not combine multiple tool calls into one message unless explicitly required.
                """.trimIndent()
            )
        )

        LLMProvider.THREED_LLM -> ProviderProfile(
            provider = provider,
            modelDefaults = ModelSettings(
                temperature = 0.7,
                topP = 0.9,
                topK = 40,
                maxTokens = 4096
            ),
            capabilities = ProviderCapabilities(
                supportsStreaming = true,
                toolCallingMode = ToolCallingMode.PROMPT_BASED,
                supportsVision = false,
                maxContextTokens = 65536
            ),
            prompt = ProviderPromptProfile(
                systemRole = null,
                toolRules = """
                    Tool rules:
                    - Your response will be parsed by an 8-stage tool-extraction pipeline.
                    - Place exactly one tool call in a fenced JSON block.
                    - Include all required arguments and no commentary inside the JSON block.
                    - If extraction fails, simplify the JSON shape and retry.
                """.trimIndent()
            ),
            sessionPolicy = SessionPolicy(
                compactionThreshold = 85,
                resetOnProviderChange = true
            )
        )
    }
}

object PromptAssembler {

    fun assemble(
        coreRules: String,
        projectContext: String? = null,
        toolProtocol: String? = null,
        profile: ProviderProfile,
        toolsEnabled: Boolean = true
    ): String = buildString {
        append(coreRules.trim())

        if (!projectContext.isNullOrBlank()) {
            append("\n\n")
            append(projectContext.trim())
        }

        if (!toolProtocol.isNullOrBlank()) {
            append("\n\n")
            append(toolProtocol.trim())
        }

        if (toolsEnabled && !profile.prompt.toolRules.isNullOrBlank()) {
            append("\n\n")
            append(profile.prompt.toolRules.trim())
        }
    }
}
